src/parallel/CollectiveCommunicationPersistent_helper.h

```diff
@@ -0,0 +1,379 @@
+
+    *pinoutvec = std::max(*pinoutvec, *pinvec);
+    
+    offset += size;
+}
+// base case
+template<typename T, typename... Ts>
+void free_max( std::byte *invec, std::byte *inoutvec, size_t size, size_t offset, T head, Ts... tail )
+{
+    T *pinvec = reinterpret_cast<T*> (invec + offset);
+    T *pinoutvec = reinterpret_cast<T*> (inoutvec + offset);
+
+    *pinoutvec = std::max(*pinoutvec, *pinvec);
+    
+    offset += size;
+    free_max(invec, inoutvec, size, offset, tail...);
+}
+// class that contains function for MPI_Op
+// this maxs the values of different types together
+struct max_struct
+{
+    template<typename... Ts>
+    static void func( void *invec, void *inoutvec, int *len, MPI_Datatype *datatype )
+    {
+        size_t size = get_max_size(Ts{}...);
+        size_t offset = 0;
+        free_max( (std::byte *) invec, (std::byte *) inoutvec, size, offset, Ts{}... );
+    }
+};
+
+
+// struct for mining
+// overload
+template<typename T>
+void free_min( std::byte *invec, std::byte *inoutvec, size_t size, size_t offset, T head )
+{
+    T *pinvec = reinterpret_cast<T*> (invec + offset);
+    T *pinoutvec = reinterpret_cast<T*> (inoutvec + offset);
+
+    *pinoutvec = std::min(*pinoutvec, *pinvec);
+    
+    offset += size;
+}
+// base case
+template<typename T, typename... Ts>
+void free_min( std::byte *invec, std::byte *inoutvec, size_t size, size_t offset, T head, Ts... tail )
+{
+    T *pinvec = reinterpret_cast<T*> (invec + offset);
+    T *pinoutvec = reinterpret_cast<T*> (inoutvec + offset);
+
+    *pinoutvec = std::min(*pinoutvec, *pinvec);
+    
+    offset += size;
+    free_min(invec, inoutvec, size, offset, tail...);
+}
+// class that contains function for MPI_Op
+// this mins the values of different types together
+struct min_struct
+{
+    template<typename... Ts>
+    static void func( void *invec, void *inoutvec, int *len, MPI_Datatype *datatype )
+    {
+        size_t size = get_max_size(Ts{}...);
+        size_t offset = 0;
+        free_min( (std::byte *) invec, (std::byte *) inoutvec, size, offset, Ts{}... );
+    }
+};
+
+
+// class to handle deallocation of MPI allocations
+class Coll_Comm_Deallocator
+{
+public:
+    static auto emplace_back(MPI_Request* request)
+    {
+        _allocated_requests.emplace_back(request);
+    }
+
+    static auto emplace_back(MPI_Op* op)
+    {
+        _allocated_ops.emplace_back(op);
+    }
+
+    static auto emplace_back(MPI_Datatype* type)
+    {
+        _allocated_types.emplace_back(type);
+    }
+
+    static void deallocate()
+    {
+        for (auto ptr : _allocated_requests)
+            MPI_Request_free(ptr);
+        for (auto ptr : _allocated_ops)
+            MPI_Op_free(ptr);
+        for (auto ptr : _allocated_types)
+            MPI_Type_free(ptr);
+    }
+
+private:
+    static inline std::vector<MPI_Request*> _allocated_requests;
+    static inline std::vector<MPI_Op*> _allocated_ops;
+    static inline std::vector<MPI_Datatype*> _allocated_types;
```

Comment:
So if I understand this correctly, these vectors accumulate all MPI requests, ops, and types created throughout the simulation. Doesn't this mean that in a very long-running simulation, this can accumulate a lot of trash if the persistent communication isn't reusing all of this perfectly? Could this lead to problems? Can we somehow build safeguards / checks against this?